nautilus_core/portfolio/src/portfolio.rs

```diff
@@ -71,138 +295,965 @@ impl Portfolio {
+            }
+        }
+
+        let mut initialized = true;
+
+        for instrument_id in instruments {
+            let positions_open: Vec<Position> = {
+                let borrowed_cache = self.cache.borrow();
+                borrowed_cache
+                    .positions_open(None, Some(&instrument_id), None, None)
+                    .into_iter()
+                    .cloned()
+                    .collect()
+            };
+
+            self.update_net_position(&instrument_id, positions_open);
+
+            let calculated_unrealized_pnl = self.calculate_unrealized_pnl(&instrument_id).unwrap();
+            let calculated_realized_pnl = self.calculate_realized_pnl(&instrument_id).unwrap();
+
+            self.unrealized_pnls
+                .insert(instrument_id, calculated_unrealized_pnl);
+            self.realized_pnls
+                .insert(instrument_id, calculated_realized_pnl);
+
+            let mut borrowed_cache = self.cache.borrow_mut();
+            let account =
+                if let Some(account) = borrowed_cache.mut_account_for_venue(&instrument_id.venue) {
+                    account
+                } else {
+                    log::error!(
+                        "Cannot update maintenance (position) margin: no account registered for {}",
+                        instrument_id.venue
+                    );
+                    initialized = false;
+                    break;
+                };
+
+            let account = match account {
+                AccountAny::Cash(_) => continue,
+                AccountAny::Margin(margin_account) => margin_account,
+            };
+
+            let borrowed_cache = self.cache.borrow_mut();
+            let instrument = if let Some(instrument) = borrowed_cache.instrument(&instrument_id) {
+                instrument
+            } else {
+                log::error!(
+                    "Cannot update maintenance (position) margin: no instrument found for {}",
+                    instrument_id
+                );
+                initialized = false;
+                break;
+            };
+
+            let result = self.accounts.update_positions(
+                account,
+                instrument.clone(),
+                self.cache
+                    .borrow()
+                    .positions_open(None, Some(&instrument_id), None, None),
+                self.clock.borrow().timestamp_ns(),
+            );
+
+            if result.is_none() {
+                initialized = false;
+            }
+        }
+
+        let open_count = all_positions_open.len();
+        self.initialized = initialized;
+        log::info!(
+            "Initialized {} open position{}",
+            open_count,
+            if open_count == 1 { "" } else { "s" }
+        );
     }
 
     pub fn update_quote_tick(&mut self, quote: &QuoteTick) {
-        todo!()
+        self.unrealized_pnls.remove(&quote.instrument_id);
```

Comment:
update quote tick uses self fields that are not behind and Rc Refcell which is why it won't work by just passing the cache or the message bus. The other update functions too use parts of self that are not cloneable references. So it seems like an inner outer style must be implemented.